Drivers: ADC: add ADC driver for Infineon Edge MCU family

jsbatch opened the pull request:
This change implements an ADC driver for the PSOC Edge family of MCUs that use the Autonomous Analog (AutAnalog) SAR ADC.

- Updates device tree and KConfig files based on the new ADC driver implementation.
- Updates modules/hal/infineon to include new support files.
- Adds AutAnalog SAR ADC driver files for PSOC Edge ADC.
- Adds sample and test overlays to provide support for:
  - samples/drivers/adc/adc_dt/
  - samples/drivers/adc/adc_sequence/
  - tests/drivers/adc/adc_api/
  - tests/drivers/adc/adc_error_cases/

jsbatch committed:
dts: infineon: pse84 device tree changes to support AutAnalog ADC

*Separates AutAnalog and AutAnalog ADC in device tree.
*Makes AutAnalog SAR ADC a child of the AutAnalog system to reflect
 hardware architecture.
*Adds binding file for AutAnalog SAR ADC driver.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
modules: hal_infineon: Adding AutAnalog files to build

*Adds AutAnalog related files to the cmakeu list and adding KConfig
 definitions for AutAnalog and AutAnalog SAR ADC.
*Updates hal_infineon in west manifest.
*Updates infineon_kconfig.h to prevent multiple declarations of defines.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
drivers: adc: Infineon AutAnalog SAR ADC driver

Adds AutAnalog SAR ADC driver files to support ADC conversions in the
PCSOC Edge family of MCUs.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
samples: adc: Adding Infineon AutAnalog ADC samples

Adding files to allow running the following samples for the Infineon
AutAnalog SAR ADC driver on the PSOC Edge family of MCUs:
  * samples/drivers/adc_dt
  * samples/drivers/adc_sequence

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
tests: adc: Adding Infineon AutAnalog ADC driver tests

Adding files to allow running the following tests for the Infineon
AutAnalog SAR ADC driver on the PSOC Edge family of MCUs:
  * tests/drivers/adc/adc_api
  * tests/drivers/adc/adc_error_cases

Signed-off-by: John Batch <john.batch@infineon.com>

JarmouniA commented on tests/drivers/adc/adc_api/boards/kit_pse84_eval_pse846gps2dbzc4a_m55.overlay:
commit title should start with `tests:.. `

jsbatch committed:
soc: infineon: Enable AutAnalog subsystem init

Changes initialization to use a function from hal_infineon.  This
initialization function will initialize the AutAnalog system if needed
as well as the SystemInit.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch commented on dts/bindings/adc/infineon,adc-base.yaml:
I've removed this file.

jsbatch commented on modules/hal_infineon/infineon_kconfig.h:
I've moved this to the modules commit.  Let me know if you think this doesn't work and needs to be separated further.  The changes are needed for a file included in the hal_infineon module in this commit.

jsbatch commented on soc/infineon/edge/pse84/soc_pse84_m55.c:
```diff
@@ -92,7 +93,7 @@ void soc_early_init_hook(void)
 
 	/* Initializes the system */
-	SystemInit();
+	ifx_cycfg_init();
```
I've moved this to a separate commit.

jsbatch commented on samples/drivers/adc/adc_dt/boards/kit_pse84_eval_pse846gps2dbzc4a_m55.overlay:
I've changed the commit title.

JarmouniA commented on dts/arm/infineon/edge/pse84/pse84.dtsi:
```diff
@@ -299,12 +299,21 @@
-		adc0: adc@42e80000 {
-			compatible = "infineon,autanalog-sar-adc";
-			reg = <0x42e80000 0xf20>;
+		autanalog: autanalog@42e00000 {
```
```suggestion
		autanalog: dsp@42e00000 {
```
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#generic-names-recommendation

jsbatch replied:
Hmm, I'm not sure dsp is really a good description.  This section of the device contains several analog components, such as ADC, DAC, comparators, etc.  It also has a simple controller to provide simple coordination between these devices, but it doesn't really do any signal processing on it's own.  

I don't really see anything in that list that I think is good fit for the device.  Perhaps it could be described as "analog" instead, as an analog subsystem?  That would at least be more generic than the original naming and I think fits with the goals they outline.

jsbatch commented on dts/bindings/adc/infineon,autanalog-sar-adc.yaml:
```diff
@@ -0,0 +1,56 @@
+      - "prb_out1"
+      - "prb_out2"
+    description: |
+      Selects which voltage reference to use for the ADC.  Note that the enum order
```
Okay, I've updated the way this binding value is used to remove the external dependency and updated the description.

ZhaoxiangJin commented on drivers/adc/adc_ifx_autanalog_sar.c:
```diff
@@ -0,0 +1,724 @@
+		/* Wait for conversion to complete */
+		do {
+			result_status = Cy_AutAnalog_SAR_GetHSchanResultStatus(0);
+		} while ((result_status & sequence->channels) != sequence->channels);
```
Yes, you are correct. Synchronization is achieved through busy-wait. Just curious, what's the maximum conversion rate for your ADC?

sreeramIfx commented on drivers/adc/adc_ifx_autanalog_sar.c:
```diff
@@ -0,0 +1,726 @@
+	struct adc_context ctx;
+	const struct device *dev;
+	/* Conversion Buffer */
+	uint16_t *buffer;
```
```suggestion
	uint16_t *conversion_buffer;
```

jsbatch replied:
Variable name has been updated as suggested.

sreeramIfx commented on drivers/adc/adc_ifx_autanalog_sar.c:
```diff
@@ -0,0 +1,726 @@
+{
+	uint8_t timer_index = IFX_AUTANALOG_SAR_SAMPLETIME_COUNT;
+	cy_stc_autanalog_sar_seq_tab_hs_t *seq_entry = &data->pdl_adc_seq_hs_cfg_obj[0];
+	/* Verify that all channels in the sequence have the same acquisition time and the sample
```
```suggestion

	/* Verify that all channels in the sequence have the same acquisition time and the sample
```

jsbatch replied:
I've added the newline.

sreeramIfx reviewed (commented):
LGTM

ZhaoxiangJin reviewed (approved):
not a block issue, just a small nit - It would be better if we could separate the format and code changes.
In addition, would you consider resolving the issues reported by the sonarqubecloud bot?
